File: server/Protocol/FrameTracePacket.cs
using System.Buffers.Binary;

namespace DeskStreamer.Server.Protocol;

/// <summary>
/// Optional per-frame timing sidecar sent after a frame's video and FEC datagrams. Keeping this
/// out of the v1 media header preserves compatibility with clients that do not understand tracing.
/// All timestamps use the server monotonic clock in microseconds. CaptureEndUs and ConvertEndUs
/// mark D3D11 command submission rather than GPU completion; measuring GPU execution itself would
/// require an asynchronously-resolved timestamp-query ring so tracing never stalls the hot path.
/// </summary>
public readonly record struct ServerFrameTrace(
    uint FrameId,
    long CaptureStartUs,
    long CaptureEndUs,
    long ConvertEndUs,
    long EncodeSubmitUs,
    long EncodeFinishUs,
    long PacketStartUs,
    long PacketEndUs);

public static class FrameTracePacket
{
    public const int Size = 68;
    public const byte Version = 1;
    private static ReadOnlySpan<byte> Magic => "DSTR"u8;

    public static void Write(Span<byte> destination, in ServerFrameTrace trace)
    {
        if (destination.Length < Size)
            throw new ArgumentException($"Frame trace destination must be at least {Size} bytes");

        Magic.CopyTo(destination);
        destination[4] = Version;
        destination[5] = destination[6] = destination[7] = 0;
        BinaryPrimitives.WriteUInt32BigEndian(destination.Slice(8, 4), trace.FrameId);
        WriteInt64(destination.Slice(12, 8), trace.CaptureStartUs);
        WriteInt64(destination.Slice(20, 8), trace.CaptureEndUs);
        WriteInt64(destination.Slice(28, 8), trace.ConvertEndUs);
        WriteInt64(destination.Slice(36, 8), trace.EncodeSubmitUs);
        WriteInt64(destination.Slice(44, 8), trace.EncodeFinishUs);
        WriteInt64(destination.Slice(52, 8), trace.PacketStartUs);
        WriteInt64(destination.Slice(60, 8), trace.PacketEndUs);
    }

    private static void WriteInt64(Span<byte> destination, long value) =>
        BinaryPrimitives.WriteInt64BigEndian(destination, value);
}

File: android/app/src/main/java/com/deskstream/client/proto/FrameTracePacket.kt
package com.deskstream.client.proto

data class ServerFrameTrace(
    val frameId: Long,
    val captureStartUs: Long,
    val captureEndUs: Long,
    val convertEndUs: Long,
    val encodeSubmitUs: Long,
    val encodeFinishUs: Long,
    val packetStartUs: Long,
    val packetEndUs: Long
)

object FrameTracePacket {
    const val SIZE = 68

    fun parse(bytes: ByteArray, length: Int): ServerFrameTrace? {
        if (length != SIZE || bytes.size < SIZE ||
            bytes[0] != 'D'.code.toByte() || bytes[1] != 'S'.code.toByte() ||
            bytes[2] != 'T'.code.toByte() || bytes[3] != 'R'.code.toByte() ||
            bytes[4].toInt() != 1 || bytes[5].toInt() != 0 ||
            bytes[6].toInt() != 0 || bytes[7].toInt() != 0
        ) return null

        return ServerFrameTrace(
            frameId = u32(bytes, 8),
            captureStartUs = i64(bytes, 12),
            captureEndUs = i64(bytes, 20),
            convertEndUs = i64(bytes, 28),
            encodeSubmitUs = i64(bytes, 36),
            encodeFinishUs = i64(bytes, 44),
            packetStartUs = i64(bytes, 52),
            packetEndUs = i64(bytes, 60)
        )
    }

    private fun u32(bytes: ByteArray, offset: Int): Long =
        ((bytes[offset].toLong() and 0xff) shl 24) or
            ((bytes[offset + 1].toLong() and 0xff) shl 16) or
            ((bytes[offset + 2].toLong() and 0xff) shl 8) or
            (bytes[offset + 3].toLong() and 0xff)

    private fun i64(bytes: ByteArray, offset: Int): Long {
        var value = 0L
        for (i in 0 until 8) value = (value shl 8) or (bytes[offset + i].toLong() and 0xff)
        return value
    }
}
